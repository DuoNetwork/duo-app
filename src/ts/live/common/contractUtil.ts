import Web3 from 'web3';
import { Contract } from 'web3/types';
import infura from '../../../../../duo-admin/src/keys/infura.json';
import custodianAbi from '../../../../../duo-admin/src/static/Custodian.json';
import duoAbi from '../../../../../duo-admin/src/static/DUO.json';
import { IAddresses, IBalances, ICustodianPrices, ICustodianStates } from '../common/types';
import * as CST from './constants';
//import util from './util';

class ContractUtil {
	private web3: Web3;
	private duo: Contract;
	private custodian: Contract;
	private isReadOnly: boolean;
	public readonly custodianAddr: string;
	private readonly duoContractAddr: string;

	constructor() {
		if (typeof (window as any).web3 !== 'undefined') {
			this.isReadOnly = false;
			this.web3 = new Web3((window as any).web3.currentProvider);
		} else {
			this.web3 = new Web3(
				new Web3.providers.HttpProvider(
					(__KOVAN__ ? CST.PROVIDER_INFURA_KOVAN : CST.PROVIDER_INFURA_MAIN) +
						'/' +
						infura.token
				)
			);
			this.isReadOnly = true;
		}
		this.custodianAddr = __KOVAN__ ? CST.CUSTODIAN_ADDR_KOVAN : CST.CUSTODIAN_ADDR_MAIN;
		this.custodian = new this.web3.eth.Contract(custodianAbi.abi, this.custodianAddr);
		this.duoContractAddr = __KOVAN__ ? CST.DUO_CONTRACT_ADDR_KOVAN : CST.DUO_CONTRACT_ADDR_MAIN;
		this.duo = new this.web3.eth.Contract(duoAbi.abi, this.duoContractAddr);
	}

	public onWeb3AccountUpdate(onUpdate: (addr: string, network: number) => any) {
		const store = (this.web3.currentProvider as any).publicConfigStore;
		if (store)
			store.on('update', () => {
				onUpdate(
					store.getState().selectedAddress || '',
					Number(store.getState().networkVersion || '')
				);
			});
	}

	public convertCustodianState(rawState: string) {
		switch (rawState) {
			case CST.STATE_INCEPTION:
				return CST.CTD_INCEPTION;
			case CST.STATE_TRADING:
				return CST.CTD_TRADING;
			case CST.STATE_PRERESET:
				return CST.CTD_PRERESET;
			case CST.STATE_UP_RESET:
				return CST.CTD_UP_RESET;
			case CST.STATE_DOWN_RESET:
				return CST.CTD_DOWN_RESET;
			case CST.STATE_PERIOD_RESET:
				return CST.CTD_PERIOD_RESET;
			default:
				return CST.CTD_LOADING;
		}
	}

	public async getSystemStates(): Promise<ICustodianStates> {
		const [states, duoBalance] = await Promise.all([
			this.custodian.methods.getSystemStates().call(),
			this.getDuoBalance(this.custodianAddr)
		]);
		return {
			state: this.convertCustodianState(states[0].valueOf()),
			navA: this.fromWei(states[1]),
			navB: this.fromWei(states[2]),
			totalSupplyA: this.fromWei(states[3]),
			totalSupplyB: this.fromWei(states[4]),
			ethBalance: this.fromWei(states[5]),
			alpha: states[6].valueOf() / 10000,
			beta: this.fromWei(states[7]),
			feeAccumulated: this.fromWei(states[8]),
			periodCoupon: this.fromWei(states[9]),
			limitPeriodic: this.fromWei(states[10]),
			limitUpper: this.fromWei(states[11]),
			limitLower: this.fromWei(states[12]),
			commissionRate: states[13] / 10000,
			period: Number(states[14].valueOf()),
			iterationGasThreshold: Number(states[15].valueOf()),
			ethDuoFeeRatio: Number(states[16].valueOf()),
			preResetWaitingBlocks: Number(states[17].valueOf()),
			priceTol: Number(states[18].valueOf() / 10000),
			priceFeedTol: Number(states[19].valueOf() / 10000),
			priceFeedTimeTol: Number(states[20].valueOf()),
			priceUpdateCoolDown: Number(states[21].valueOf()),
			numOfPrices: Number(states[22].valueOf()),
			nextResetAddrIndex: Number(states[23].valueOf()),
			lastAdminTime: Number(states[24].valueOf()),
			adminCoolDown: Number(states[25]),
			usersLength: Number(states[26].valueOf()),
			addrPoolLength: Number(states[27].valueOf()),
			duoBalance: duoBalance
		};
	}

	public async getSystemAddresses(): Promise<IAddresses> {
		const addr: string[] = await this.custodian.methods.getSystemAddresses().call();
		const balances = await Promise.all(addr.map(a => this.getEthBalance(a)));
		return {
			operator: {
				address: addr[0],
				balance: balances[0]
			},
			feeCollector: {
				address: addr[1],
				balance: balances[1]
			},
			priceFeed1: {
				address: addr[2],
				balance: balances[2]
			},
			priceFeed2: {
				address: addr[3],
				balance: balances[3]
			},
			priceFeed3: {
				address: addr[4],
				balance: balances[4]
			},
			poolManager: {
				address: addr[5],
				balance: balances[5]
			}
		};
	}

	public async getSystemPrices(): Promise<ICustodianPrices> {
		const prices = await this.custodian.methods.getSystemPrices().call();
		const custodianPrices = [0, 1, 2, 3].map(i => ({
			address: prices[i * 3].valueOf(),
			price: this.fromWei(prices[1 + i * 3]),
			timestamp: prices[2 + i * 3].valueOf() * 1000
		}));

		return {
			first: custodianPrices[0],
			second: custodianPrices[1],
			reset: custodianPrices[2],
			last: custodianPrices[3]
		};
	}

	public async getBalances(address: string): Promise<IBalances> {
		if (!address)
			return {
				eth: 0,
				duo: 0,
				allowance: 0,
				tokenA: 0,
				tokenB: 0
			};

		const balances = await Promise.all([
			this.getEthBalance(address),
			this.getDuoBalance(address),
			this.getDuoAllowance(address),
			this.getTokenBalance(address, true),
			this.getTokenBalance(address, false)
		]);

		return {
			eth: balances[0],
			duo: balances[1],
			allowance: balances[2],
			tokenA: balances[3],
			tokenB: balances[4]
		};
	}

	public getUserAddress(index: number) {
		return this.custodian.methods.users(index).call();
	}

	public getPoolAddress(index: number) {
		return this.custodian.methods.addrPool(index).call();
	}

	public async getGasPrice(): Promise<number> {
		return await this.web3.eth.getGasPrice();
	}

	public async getCurrentAddress(): Promise<string> {
		return (await this.web3.eth.getAccounts())[0];
	}

	public getCurrentNetwork(): Promise<number> {
		return this.web3.eth.net.getId();
	}

	public async getEthBalance(address: string): Promise<number> {
		return this.fromWei(await this.web3.eth.getBalance(address));
	}

	private async getDuoBalance(address: string): Promise<number> {
		return this.fromWei(await this.duo.methods.balanceOf(address).call());
	}

	private async getDuoAllowance(address: string): Promise<number> {
		return this.fromWei(await this.duo.methods.allowance(address, this.custodianAddr).call());
	}

	private async getTokenBalance(address: string, isA: boolean): Promise<number> {
		return this.fromWei(await this.custodian.methods.balanceOf(isA ? 0 : 1, address).call());
	}

	public fromWei(value: string | number) {
		return Number(this.web3.utils.fromWei(value, 'ether'));
	}

	public toWei(value: string | number) {
		return this.web3.utils.toWei(value + '', 'ether');
	}

	public checkAddress(addr: string) {
		if (!addr.startsWith('0x') || addr.length !== 42) return false;
		return this.web3.utils.checkAddressChecksum(addr);
	}

	public create(
		address: string,
		value: number,
		payFeeInEth: boolean,
		onTxHash: (hash: string) => any
	) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		return this.custodian.methods
			.create(payFeeInEth)
			.send({
				from: address,
				value: this.toWei(value)
			})
			.on('transactionHash', onTxHash);
	}

	public redeem(
		address: string,
		amtA: number,
		amtB: number,
		payFeeInEth: boolean,
		onTxHash: (hash: string) => any
	) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		return this.custodian.methods
			.redeem(this.toWei(amtA), this.toWei(amtB), payFeeInEth)
			.send({
				from: address
			})
			.on('transactionHash', onTxHash);
	}

	public duoApprove(address: string, spender: string, value: number) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		return this.duo.methods.approve(spender, this.toWei(value)).send({
			from: address
		});
	}

	public duoTransfer(address: string, to: string, value: number) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		return this.duo.methods.transfer(to, this.toWei(value)).send({
			from: address
		});
	}

	public approve(address: string, spender: string, value: number, isA: boolean) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		return this.custodian.methods.approve(isA ? 0 : 1, spender, this.toWei(value)).send({
			from: address
		});
	}

	public transfer(address: string, to: string, value: number, isA: boolean) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');

		// dummy from address
		return this.custodian.methods.transfer(isA ? 0 : 1, address, to, this.toWei(value)).send({
			from: address
		});
	}

	public collectFee(address: string, amount: number) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');
		return this.custodian.methods.collectFee(this.toWei(amount)).send({
			from: address
		});
	}

	public setValue(address: string, index: number, newValue: number) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');
		return this.custodian.methods.setValue(index, newValue).send({
			from: address
		});
	}

	public addAddress(address: string, addr1: string, addr2: string) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');
		return this.custodian.methods.addAddress(addr1, addr2).send({
			from: address
		});
	}

	public removeAddress(address: string, addr: string) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');
		return this.custodian.methods.addAddress(addr).send({
			from: address
		});
	}

	public updateAddress(address: string, currentRole: string) {
		if (this.isReadOnly) return Promise.reject('Read Only Mode');
		return this.custodian.methods.addAddress(currentRole).send({
			from: address
		});
	}
}

const contractUtil = new ContractUtil();
export default contractUtil;
