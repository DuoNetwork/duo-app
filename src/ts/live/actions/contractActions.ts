import * as CST from '../common/constants';
import contractUtil from '../common/contractUtil';
import * as reduxTypes from '../common/reduxTypes';
import { ICustodianPrice, ICustodianStates } from '../common/types';

export function custodianStatesUpdate(states: ICustodianStates): reduxTypes.Action {
	return {
		type: CST.AC_CTD_STATES,
		value: states
	};
}

export function getCustodianStates(): reduxTypes.ThunkAction {
	return async dispatch => {
		const states = await contractUtil.getSystemStates();
		dispatch(custodianStatesUpdate(states));
	};
}

export function custodianPricesUpdate(prices: {
	[name: string]: ICustodianPrice;
}): reduxTypes.Action {
	return {
		type: CST.AC_CTD_PRICES,
		value: prices
	};
}

export function getCustodianPrices(): reduxTypes.ThunkAction {
	return async dispatch => {
		const prices = await contractUtil.getSystemPrices();
		dispatch(
			custodianPricesUpdate({
				reset: prices[0],
				last: prices[1]
			})
		);
	};
}
