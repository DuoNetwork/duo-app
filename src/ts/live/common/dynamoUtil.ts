import AWS from 'aws-sdk';
import { QueryInput, QueryOutput, ScanInput, ScanOutput } from 'aws-sdk/clients/dynamodb';
import devConfig from '../../keys/aws.ui.dev.json';
import liveConfig from '../../keys/aws.ui.live.json';
import { IPriceBar, IPriceStatus, IStatus } from '../common/types';
import * as CST from './constants';

export class DynamoUtil {
	private ddb: AWS.DynamoDB;
	constructor() {
		// this.role = role;
		AWS.config.update(__DEV__ ? devConfig : liveConfig);
		this.ddb = new AWS.DynamoDB({ apiVersion: CST.AWS_DYNAMO_API_VERSION });
	}

	public scanData(params: ScanInput): Promise<ScanOutput> {
		return new Promise((resolve, reject) => {
			this.ddb.scan(params, (err, data) => (err ? reject(err) : resolve(data)));
		});
	}

	public queryData(params: QueryInput): Promise<QueryOutput> {
		return new Promise((resolve, reject) =>
			this.ddb.query(params, (err, data) => (err ? reject(err) : resolve(data)))
		);
	}

	public async queryHourlyOHLC(source: string, dates: string[]) {
		const promiseList = dates.map(date => this.queryData({
				TableName: __DEV__ ? CST.DB_AWS_HOURLY_DEV : CST.DB_AWS_HOURLY_LIVE,
				KeyConditionExpression: CST.DB_HR_SRC_DATE + ' = :' + CST.DB_HR_SRC_DATE,
				ExpressionAttributeValues: {
					[':' + CST.DB_HR_SRC_DATE]: { S: source + '|' + date }
				}
			}));

		const allData: IPriceBar[] = [];
		(await Promise.all(promiseList)).forEach(r => allData.push(...this.convertHourly(r)));
		allData.sort((a, b) => a.timestamp - b.timestamp);
		return allData;
	}

	public convertHourly(hourly: QueryOutput): IPriceBar[] {
		if (!hourly.Items || !hourly.Items.length) return [];
		const sourceDate = hourly.Items[0][CST.DB_HR_SRC_DATE].S || '';
		const [source, date] = sourceDate.split('|');
		return hourly.Items.map(h => {
			const hour = h[CST.DB_HR_HOUR].N || '';
			return {
				source: source,
				date: date,
				hour: hour.length < 2 ? '0' + hour : hour,
				minute: 0,
				open: Number(h[CST.DB_OHLC_OPEN].N),
				high: Number(h[CST.DB_OHLC_HIGH].N),
				low: Number(h[CST.DB_OHLC_LOW].N),
				close: Number(h[CST.DB_OHLC_CLOSE].N),
				volume: Number(h[CST.DB_OHLC_VOLUME].N),
				timestamp: Number(h[CST.DB_OHLC_TS].N)
			};
		});
	}

	public async scanStatus() {
		return this.convertStatus(
			await this.scanData({
				TableName: __DEV__ ? CST.DB_AWS_STATUS_DEV : CST.DB_AWS_STATUS_LIVE
			})
		);
	}

	public convertStatus(status: ScanOutput): IStatus[] {
		if (!status.Items || !status.Items.length) return [];

		const output = status.Items.map(d => {
			const process = d[CST.DB_ST_PROCESS].S || '';
			const timestamp = Number(d[CST.DB_ST_TS].N);
			if (process.startsWith('PRICE'))
				return {
					process: process,
					timestamp: timestamp,
					price: Number(d[CST.DB_TX_PRICE].N),
					volume: Number(d[CST.DB_TX_AMOUNT].N)
				} as IPriceStatus;
			else
				return {
					process: process,
					timestamp: timestamp
				};
		});
		output.sort((a, b) => b.timestamp - a.timestamp);
		output.sort((a, b) => a.process.localeCompare(b.process));

		return output;
	}
}

const dynamoUtil = new DynamoUtil();
export default dynamoUtil;
