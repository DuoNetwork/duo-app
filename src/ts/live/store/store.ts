import { AnyAction, applyMiddleware, createStore } from 'redux';
import { createLogger } from 'redux-logger';
import thunk, { ThunkDispatch, ThunkMiddleware } from 'redux-thunk';
import { IState } from '../common/types';
import reducers from '../reducers';

const middleWares = [thunk as ThunkMiddleware<IState, AnyAction>];
if (__DEV__) middleWares.push(createLogger());

const store = createStore(
	reducers,
	{},
	applyMiddleware<ThunkDispatch<IState, undefined, AnyAction>, IState>(...middleWares)
);

export default store;
