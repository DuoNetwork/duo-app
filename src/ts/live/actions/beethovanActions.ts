import moment from 'moment';
import dynamoUtil from '../../../../../duo-admin/src/utils/dynamoUtil';
import chartUtil from '../common/chartUtil';
import * as CST from '../common/constants';
import contract from '../common/contract';
import {
	IAcceptedPrice,
	IAccountBalances,
	IAddress,
	IBeethovanAddresses,
	IBeethovanBalances,
	IBeethovanPrices,
	IBeethovanStates,
	IConversion,
	IPrice,
	VoidThunkAction
} from '../common/types';
import util from '../common/util';
import * as web3Actions from './web3Actions';

export function statesUpdate(states: IBeethovanStates) {
	return {
		type: CST.AC_BTV_STATES,
		value: states
	};
}

export function getStates(): VoidThunkAction {
	return async dispatch => {
		const states = await contract.getCustodianStates();
		dispatch(statesUpdate(states));
	};
}

export function pricesUpdate(prices: IBeethovanPrices) {
	return {
		type: CST.AC_BTV_PRICES,
		value: prices
	};
}

export function getPrices(): VoidThunkAction {
	return async dispatch => dispatch(pricesUpdate(await contract.getCustodianPrices()));
}

export function balancesUpdate(balance: IBeethovanBalances) {
	return {
		type: CST.AC_BTV_BALANCES,
		value: balance
	};
}

export function getBalances(): VoidThunkAction {
	return async (dispatch, getState) =>
		dispatch(balancesUpdate(await contract.getBalances(getState().web3.account)));
}

export function allBalancesUpdate(balance: IAccountBalances, index: number) {
	return {
		type: CST.AC_ALL_BALANCES,
		value: {
			[index]: balance
		}
	};
}

export function getAllBalances(start: number, end: number): VoidThunkAction {
	return async dispatch => {
		for (let i = start; i < end; i++)
			contract.getUserAddress(i).then((account: any) => {
				if (account)
					contract.getBalances(account).then(balance =>
						dispatch(
							allBalancesUpdate(
								{
									account: account,
									...balance
								},
								i
							)
						)
					);
			});
	};
}

export function addressesUpdate(addr: IBeethovanAddresses) {
	return {
		type: CST.AC_BTV_ADDRESSES,
		value: addr
	};
}

export function addressPoolUpdate(address: IAddress[]) {
	return {
		type: CST.AC_ADDR_POOL,
		value: address
	};
}

export function getAddresses(): VoidThunkAction {
	return async (dispatch, getState) => {
		dispatch(addressesUpdate(await contract.getCustodianAddresses()));
		const poolLength = getState().beethovan.beethovanStates.addrPoolLength;
		const addrPool: IAddress[] = [];
		for (let i = 0; i < poolLength; i++) {
			const address = await contract.getPoolAddress(i);
			if (address)
				addrPool.push({
					address: address,
					balance: await contract.getEthBalance(address)
				});
		}
		dispatch(addressPoolUpdate(addrPool));
	};
}

export function exchangePricesUpdate(prices: IPrice[]) {
	return {
		type: CST.AC_BTV_EX_PX,
		value: prices
	};
}

export function fetchExchangePrices(): VoidThunkAction {
	return async (dispatch, state) => {
		const source = state().ui.source;
		const period = state().ui.period;
		const start =
			period === 60
				? util.getUTCNowTimestamp() - period * 96 * 60000
				: util.getUTCNowTimestamp() - 400 * 60000;
		dispatch(
			exchangePricesUpdate(
				await dynamoUtil.getPrices(source, period === 60 ? 60 : 1, start, 0, 'ETH|USD')
			)
		);
	};
}

export function acceptedPricesUpdate(acceptedPrices: IAcceptedPrice[]) {
	return {
		type: CST.AC_BTV_ACCEPTED_PX,
		value: acceptedPrices
	};
}

export function fetchAcceptedPrices(contractAddress: string): VoidThunkAction {
	return async (dispatch, getState) => {
		const dates = util.getDates(4, 1, 'day', 'YYYY-MM-DD');
		const priceData = await dynamoUtil.queryAcceptPriceEvent(contractAddress, dates);
		const states = getState().beethovan.beethovanStates;
		dispatch(
			acceptedPricesUpdate(
				chartUtil.mergeReset(
					priceData,
					chartUtil.reset(
						priceData,
						states.limitUpper,
						states.limitLower,
						states.limitPeriodic
					)
				)
			)
		);
	};
}

export function conversionsUpdate(conversions: IConversion[]) {
	return {
		type: CST.AC_BTV_CONVERSIONS,
		value: conversions
	};
}

export function fetchConversions(contractAddress: string): VoidThunkAction {
	return async (dispatch, getState) => {
		const nowTimestamp = util.getUTCNowTimestamp();
		const dates = util.getDates(7, 1, 'day', 'YYYY-MM-DD');
		const account = getState().web3.account;
		const conversion = await dynamoUtil.queryConversionEvent(contractAddress, account, dates);
		(await dynamoUtil.queryUIConversionEvent(contractAddress, account)).forEach(uc => {
			if (
				!dates.includes(moment.utc(uc.timestamp).format('YYYY-MM-DD')) ||
				conversion.some(c => c.transactionHash === uc.transactionHash) ||
				(uc.pending && nowTimestamp - uc.timestamp > CST.PENDING_TX_TIMEOUT)
			)
				dynamoUtil.deleteUIConversionEvent(account, uc);
			else conversion.push(uc);
		});
		conversion.sort((a, b) => -a.timestamp + b.timestamp);
		dispatch(conversionsUpdate(conversion));
	};
}

export function refresh(): VoidThunkAction {
	return async dispatch => {
		await dispatch(getStates());
		dispatch(getPrices());
		dispatch(getBalances());
		dispatch(getAddresses());
		dispatch(fetchExchangePrices());
		dispatch(fetchAcceptedPrices(contract.custodianAddr));
		dispatch(fetchConversions(contract.custodianAddr));
		window.setInterval(async () => {
			await dispatch(getStates());
			dispatch(getPrices());
			dispatch(getBalances());
			dispatch(getAddresses());
			dispatch(fetchExchangePrices());
			dispatch(fetchAcceptedPrices(contract.custodianAddr));
			dispatch(fetchConversions(contract.custodianAddr));
		}, 60000);
	};
}

export function adminActions(): VoidThunkAction {
	return async dispatch => {
		await dispatch(getStates());
		dispatch(getAddresses());
		dispatch(web3Actions.getAccount());
		window.setInterval(async () => {
			await dispatch(getStates());
			dispatch(getAddresses());
			dispatch(web3Actions.getAccount());
		}, 60000);
	};
}

export function userActions(start: number, end: number): VoidThunkAction {
	return async dispatch => {
		await dispatch(getStates());
		dispatch(getAllBalances(start, end));
		window.setInterval(async () => {
			await dispatch(getStates());
			dispatch(getAllBalances(start, end));
		}, 60000);
	};
}

export function statusActions(): VoidThunkAction {
	return async dispatch => {
		await dispatch(getStates());
		window.setInterval(async () => {
			await dispatch(getStates());
		}, 60000);
	};
}
