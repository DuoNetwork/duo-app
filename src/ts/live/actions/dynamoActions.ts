import moment from 'moment';
import chartUtil from '../common/chartUtil';
import * as CST from '../common/constants';
import dynamoUtil from '../common/dynamoUtil';
import * as reduxTypes from '../common/reduxTypes';
import { IPriceBars } from '../common/types';

export function statusUpdate(status: object): reduxTypes.Action {
	return {
		type: CST.AC_DNM_STATUS,
		value: status
	};
}

export function scanStatus(): reduxTypes.ThunkAction {
	return async dispatch => {
		const states = await dynamoUtil.scanStatus();
		dispatch(statusUpdate(states));
	};
}

export function hourlyUpdate(hourly: IPriceBars): reduxTypes.Action {
	return {
		type: CST.AC_DMN_HOURLY,
		value: hourly
	};
}

export function fetchHourly(): reduxTypes.ThunkAction {
	return async dispatch => {
		const dates: string[] = [];
		const date = moment.utc();
		for (let i = 0; i < 7; i++) {
			dates.push(date.format('YYYY-MM-DD'))
			date.subtract(1, 'day');
		}
		const promistList = CST.EXCHANGES.map(src => dynamoUtil.queryHourlyOHLC(src, dates));
		const results = await Promise.all(promistList);
		const hourly: IPriceBars = {
			bitfinex: [],
			gemini: [],
			kraken: [],
			gdax: []
		};
		results.forEach((r, i) => (hourly[CST.EXCHANGES[i].toLowerCase()] = chartUtil.interpolate(r, true)));
		dispatch(hourlyUpdate(hourly));
	};
}

export function minutelyUpdate(minutely: IPriceBars): reduxTypes.Action {
	return {
		type: CST.AC_DMN_MINUTELY,
		value: minutely
	};
}

export function fetchMinutely(): reduxTypes.ThunkAction {
	return async dispatch => {
		const dates: string[] = [];
		const date = moment.utc();
		for (let i = 0; i < 2; i++) {
			dates.push(date.format('YYYY-MM-DD-HH'))
			date.subtract(1, 'hour');
		}
		const promistList = CST.EXCHANGES.map(src => dynamoUtil.queryMinutelyOHLC(src, dates));
		const results = await Promise.all(promistList);
		const minutely: IPriceBars = {
			bitfinex: [],
			gemini: [],
			kraken: [],
			gdax: []
		};
		results.forEach((r, i) => (minutely[CST.EXCHANGES[i].toLowerCase()] = chartUtil.interpolate(r, false)));
		dispatch(minutelyUpdate(minutely));
	};
}
