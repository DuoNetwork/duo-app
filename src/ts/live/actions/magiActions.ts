import dynamoUtil from '../../../../../duo-admin/src/utils/dynamoUtil';
import * as CST from '../common/constants';
import { IAcceptedPrice, VoidThunkAction } from '../common/types';
import util from '../common/util';

export function acceptedPricesUpdate(acceptedPrices: IAcceptedPrice[]) {
	return {
		type: CST.AC_MAG_ACCEPTED_PX,
		value: acceptedPrices
	};
}

export function fetchAcceptedPrices(contractAddress: string): VoidThunkAction {
	return async dispatch =>
		dispatch(
			acceptedPricesUpdate(
				(await dynamoUtil.queryAcceptPriceEvent(
					contractAddress,
					util.getDates(7, 1, 'day', 'YYYY-MM-DD')
				)).sort((a, b) => - a.timestamp + b.timestamp)
			)
		);
}

export function subscriptionUpdate(intervalId: number) {
	return {
		type: CST.AC_MAG_SUB,
		value: intervalId
	};
}

export function refresh(contractAddress: string): VoidThunkAction {
	return async dispatch => dispatch(fetchAcceptedPrices(contractAddress));
}

export function subscribe(contractAddress: string): VoidThunkAction {
	return async dispatch => {
		dispatch(subscriptionUpdate(0));
		dispatch(refresh(contractAddress));
		dispatch(
			subscriptionUpdate(
				window.setInterval(() => dispatch(refresh(contractAddress)), 1800000)
			)
		);
	};
}
