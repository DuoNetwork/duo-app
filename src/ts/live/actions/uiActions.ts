import * as CST from '../common/constants';
import * as reduxTypes from '../common/reduxTypes';
import * as contractActions from './contractActions';
import * as dynamoActions from './dynamoActions';

export function refreshUpdate(): reduxTypes.Action {
	return {
		type: CST.AC_REFRESH
	};
}
export function refresh(): reduxTypes.ThunkAction {
	return async dispatch => {
		dispatch(contractActions.getCustodianStates());
		dispatch(contractActions.getCustodianPrices());
		dispatch(contractActions.getBalances());
		dispatch(dynamoActions.fetchHourly());
		dispatch(dynamoActions.fetchMinutely());
		dispatch(refreshUpdate());
	};
}
