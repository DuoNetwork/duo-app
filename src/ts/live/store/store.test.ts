//import * as CST from '../common/constants';
import * as contractActions from '../actions/contractActions';
import * as dynamoActions from '../actions/dynamoActions';
import contractUtil from '../common/contractUtil';
import dynamoUtil from '../common/dynamoUtil';
import store from './store';

describe('store', () => {
	test('actions', () => {
		contractUtil.getSystemStates = jest.fn(() =>
			Promise.resolve({
				test: 'test'
			})
		);
		contractUtil.getSystemPrices = jest.fn(() => Promise.resolve(['reset', 'last']));
		contractUtil.getBalances = jest.fn(() =>
			Promise.resolve({
				test: 'test'
			})
		);
		contractUtil.getSystemAddresses = jest.fn(() =>
			Promise.resolve({
				test: 'test'
			})
		);
		dynamoUtil.scanStatus = jest.fn(() =>
			Promise.resolve({
				test: 'test'
			})
		);
		dynamoUtil.queryHourlyOHLC = jest.fn(() =>
			Promise.resolve({
				test: 'test'
			})
		);
		store.dispatch(contractActions.getCustodianStates());
		store.dispatch(contractActions.getCustodianPrices());
		store.dispatch(contractActions.getBalances());
		store.dispatch(contractActions.getAddresses());
		store.dispatch(dynamoActions.scanStatus());
		store.dispatch(dynamoActions.fetchHourly());
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getState()).toMatchSnapshot();
				resolve();
			}, 1000)
		);
	});
});
