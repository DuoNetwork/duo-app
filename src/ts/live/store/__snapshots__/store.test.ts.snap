// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store actions 1`] = `
Object {
  "contract": Object {
    "addresses": Object {
      "test": "test",
    },
    "balances": Object {
      "test": "test",
    },
    "prices": Array [
      "reset",
      "last",
    ],
    "states": Object {
      "test": "test",
    },
  },
  "dynamo": Object {
    "hourly": Object {
      "bitfinex": Object {
        "test": "test",
      },
      "gdax": Object {
        "test": "test",
      },
      "gemini": Object {
        "test": "test",
      },
      "kraken": Object {
        "test": "test",
      },
    },
    "status": Object {
      "test": "test",
    },
  },
}
`;
